Primavera enhancements/fixes

brandonherzog opened the pull request:
Hi Jon, as a frequent user of your library I've been maintaining a sizeable set of changes that enhances MPXJ's Primavera functionality and fixes some bugs that were necessary for a tool I developed. The primary focus has been on reading from XER files; however, the database reader should equally benefit from the changes.

Opening this pull request to see if you have any interest in integrating the changes into the main repository.

I'm also happy to better annotate the individual changes made.

joniles commented:
I've been following your fork for a while, and I'm very keen to incorporate your changes. I'll go through the diff and make some comments on the changes to get us started.

joniles commented on src/net/sf/mpxj/primavera/ExternalPredecessorRelation.java:
```diff
@@ -0,0 +1,132 @@
+ * This class represents the relationship between two tasks in Primavera when the predecessor
+ * task is not contained in the same project.
+ */
+public final class ExternalPredecessorRelation
```
I wonder if we could extend the existing `Relation` class here, and call the new class `ExternalRelation`. We'd set the `sourceTask` to `null`, but provide the `sourceUniqueID` attribute as you have.

brandonherzog replied:
I did struggle a bit on how to best handle cross-project task links without altering the core MPXJ MS Project model. I wasn't sure if there were any big cons to linking the tasks in the final Relation objects (such that `getSourceTask()` and `getTargetTask()` return the actual Task objects from different projects). It is certainly more convenient in my client code for this to be the case, but I could see an argument that client code should be responsible for finding the concrete Task as you suggest.

How were you thinking client code would work with an ExternalRelation object that extends Relation? It seems it could get messy in code that expects to handle files agnostic of whether they are MS Project, Primavera, etc.

```java
Task task = ...;
for (Relation relation : task.getPredecessors()) {
	Task pred = relation.getSourceTask();
	// this seems to force less than pleasant code
	if (pred == null) {
		try {
			ExternalRelation extRelation = (ExternalRelation) relation;
			pred = lookupRealTaskSomewhere(extRelation.getSourceUniqueId());
		} catch (ClassCastException ex){
			// oops?!
		}
	}
	
	// use pred
	...
}
```

joniles replied:
It was really just the class definition I was referring to, but looking at it again, let's just leave things as they are here.

joniles commented on src/net/sf/mpxj/primavera/PrimaveraDatabaseReader.java:
```diff
@@ -264,8 +267,7 @@ private void processTasks() throws SQLException
    {
       List<Row> wbs = getRows("select * from " + m_schema + "projwbs where proj_id=? and delete_date is null order by parent_wbs_id,seq_num", m_projectID);
       List<Row> tasks = getRows("select * from " + m_schema + "task where proj_id=? and delete_date is null", m_projectID);
-      List<Row> costs = getRows("select * from " + m_schema + "projcost where proj_id=? and delete_date is null", m_projectID);
```
What was the issue with `costs`?

brandonherzog replied:
XER files don't seem to actually contain the PROJCOST table (view?) so I had to build up the cost information myself instead of reading from PROJCOST in order to also support XER file exports.

joniles commented on src/net/sf/mpxj/primavera/PrimaveraDatabaseReader.java:
```diff
@@ -526,14 +528,36 @@ public String getSchema()
+   }
+
+   /**
+    * Override the default field name mapping for Resource user defined types.
```
`Resource` -> `Assignment`

brandonherzog replied:
Fixed: 5141b71

joniles commented on src/net/sf/mpxj/primavera/PrimaveraDatabaseReader.java:
```diff
@@ -526,14 +528,36 @@ public String getSchema()
     *
     * @param type target user defined data type
-    * @param fieldName field name
+    * @param fieldName field names
```
`names` -> `fieldNames`

brandonherzog replied:
Fixed: 5141b71

joniles commented on src/net/sf/mpxj/primavera/PrimaveraPMFileReader.java:
```diff
@@ -639,42 +639,45 @@ private void updateDates(Task parentTask)
             }
 
-            if (actualFinishDate == null || DateHelper.compare(actualFinishDate, task.getActualFinish()) < 0)
+            if (actualFinishDate == null || (task.getActualFinish() != null && DateHelper.compare(actualFinishDate, task.getActualFinish()) < 0))
```
I'm not sure this is necessary as `DateHelper.compare(null, null)` will be `0`.
Same comment for the rest of the changes in this file.

brandonherzog replied:
Yes, you are correct that the null check isn't necessary for start dates and I can eliminate the unnecessary null check for start dates. However, it is actually needed for finish dates because `DateHelper.compare(actualDate, null) < 0 == true` so you can wind up incorrectly replacing valid finish dates with null unless you add the extra null check. Doing a quick code search, the same subtle bug appears to exist in the Phoenix reader code (if dates can be null from Phoenix). 

Whether you define null to be less than or greater than object instances, you're going to have an issue for either start or finish; do you think it's worth adding a null-safe `DateHelp.min/max` to fix and simplify this sort of code?

joniles replied:
I like the idea of adding a null-safe `DateHelper.min/max`. I'm sure there are other places in the code we could use it. I was also wondering about factoring out the `updateDates` methods from the `PrimaveraPMFileReader` and `PrimaveraReader` classes into a single implementation in a helper class. I think we can just use the code from `PrimaveraReader`.

brandonherzog replied:
Ok, went ahead and added the null-safe `DateHelper.min/max` which simplifies the date rollup code a lot and makes sure that bug is fixed. However, the three places I found similar date rollup logic were all rolling up slightly different sets of dates so I didn't factor out the `updateDates` functions into a single common implementation.

joniles commented on src/net/sf/mpxj/primavera/PrimaveraReader.java:
```diff
@@ -141,6 +161,32 @@ public void processProjectProperties(List<Row> rows)
+      {
+         Row row = rows.get(0);
+
+         Map<String, Object> customProperties = m_project.getProjectProperties().getCustomProperties();
```
You can simply create the `Map` here and pass it to `setCustomProperties`. The reader is responsible for populating all of the attributes, so this value won't have been set elsewhere.

brandonherzog replied:
Ok, changed: db9991d

joniles commented on src/net/sf/mpxj/primavera/PrimaveraReader.java:
```diff
@@ -495,10 +663,12 @@ public void processTasks(List<Row> wbs, List<Row> tasks, List<Row> costs, List<R
          {
             task = parentTask.addTask();
          }
+         task.setProject(m_project.getProjectProperties().getName()); // P6 task always belongs to project
```
Not sure this is worth populating, unless you actually need it? It feels like a very "MS Project" thing?

brandonherzog replied:
It is; however, at least for myself I do actually use this field in client code that reads from both Project and Primavera and it's convenient for it to behave equivalently. At minimum it seems correct, if potentially superfluous.

joniles replied:
OK, makes sense. Let's leave it in.

joniles commented on src/net/sf/mpxj/primavera/PrimaveraReader.java:
```diff
@@ -495,10 +663,12 @@ public void processTasks(List<Row> wbs, List<Row> tasks, List<Row> costs, List<R
          processFields(m_taskFields, row, task);
 
          task.setMilestone(BooleanHelper.getBoolean(MILESTONE_MAP.get(row.getString("task_type"))));
+         task.setIgnoreResourceCalendar("TT_Rsrc".equals(row.getString("task_type")));
```
Could you add a comment above this line to explain the logic here?

brandonherzog replied:
Good catch, seem to be missing the `!` to negate the condition. Fixed: 2842576

joniles commented on src/net/sf/mpxj/primavera/PrimaveraReader.java:
```diff
@@ -692,33 +863,39 @@ private void addTaskUDFValue(Task task, Row row)
 
+            case BOOLEAN:
+            {
+               value = STATICTYPE_UDF_MAP.get(row.getString("udf_text"));
```
I'd prefer to leave the original values here, that way the application using the library can choose how to interpret them.

brandonherzog replied:
I understand that. However, being selfish for a second, my specific use case is hurt by lumping text fields and P6 indicator fields together. I need to be able to allow users to select a "text" field. Users did not like having tons of indicator fields mixed in containing "nonsense" values (and the indicator fields cut into the max text field count imposed by the core MPXJ MS Project model). My client code could maybe scan each text column for value that look like indicators and try to eliminate them, but that's definitely a kludge.

Any thoughts on how to make everyone happy?

joniles replied:
I wonder if we can pull out the UDF handling into a separate "manager" class, maybe to replace or augment the existing `UserFieldCounters` class. That way everyone gets the default implementation, and you can provide something more suited to your use case when you instantiated the reader.

brandonherzog replied:
Hmm... perhaps. I guess all that's needed to suite my use case but leave the default alone is to add a `setMapIndicatorUdfToFlag` sort of function to the reader classes to toggle the behavior, but I wonder if there's something more flexible that makes sense and isn't a large code lift. I'll have to think about this one a little bit more.

joniles replied:
Happy with either approach!

brandonherzog replied:
This actually had a simple solution. I switched the default back to mapping the P6 static type indicators to Text so the default behavior stays the same as before. I just left in the support I added for parsing the indicator text as booleans and now in my specific client code I can use the pre-existing `PrimaveraXERFileReader.setFieldNamesForTaskUdfType` function to set my preference for Flag. I think that should make everyone happy. eb585e68

joniles commented on src/net/sf/mpxj/primavera/UserFieldCounters.java:
```diff
@@ -48,36 +52,49 @@ public UserFieldCounters()
+      }
+
+      // no more fields available
+      throw new IllegalArgumentException("No fields for type " + type + " available");
```
We had code similar to this before. Where we had data from P6 with large numbers of UDFs we'd trigger this error. The preferred behaviour is to simply stop reading this type of UDF. Most people aren't interested in UDF values, so throwing an exception here means that they can't work with the data that they are interested in, which we managed to read OK.

brandonherzog replied:
I think I may have filed that original bug report :) The exception throw here gets caught and swallowed by the try catch in `addUserDefinedField` and that field is just dropped as you mention.

Some of my users had 100+ user defined fields and complained about the 30 field limit; thus why I expanded capacity by filling into the enterprise fields. I couldn't come up with a better way to expand capacity without altering the core MPXJ MS Project model.

joniles replied:
Got it - thanks.

brandonherzog commented:
Great; look forward to trying to get everything merged back together! Responded to all your comments and pushed commits for all the easy fixes.

Sounds like the larger things that need some more thought and/or coming to a consensus on are:
- best way to deal with cross project activity relations
- best way to handle P6 indicator custom fields

joniles commented on src/net/sf/mpxj/primavera/PrimaveraReader.java:
```diff
@@ -416,29 +512,93 @@ private ProjectCalendar getResourceCalendar(Integer calendarID)
+         {
+            Row nextRow = rows.get(i + 1);
+            Integer nextResourceID = nextRow.getInteger("rsrc_id");
+            if (resourceID == nextResourceID)
```
I'm getting a warning here about a suspicious integer comparison. Perhaps this instead:
```
            int nextResourceID = nextRow.getInt("rsrc_id");
            if (resourceID.intValue() == nextResourceID)
```

brandonherzog replied:
Fixed: f131121

joniles commented on src/net/sf/mpxj/primavera/PrimaveraReader.java:
```diff
@@ -416,29 +512,93 @@ private ProjectCalendar getResourceCalendar(Integer calendarID)
+   public void processResourceRates(List<Row> rows)
+   {
+      // Primavera defines resource cost tables by start dates so sort and define end by next
+      rows.sort(new Comparator<Row>()
```
I'm missing a `sort` method on `List<T>`.. perhaps it's a target JDK version (we need to aim for 1.6). Try this:
```java
      Collections.sort(rows, new Comparator<Row>()
      {
```

brandonherzog replied:
Yes, was targeting a later version, fixed for Java 6: d21fcaf

joniles commented:
Thanks for making the various changes. Are you happy with the code as it now stands?

joniles commented:
Also... would you be able to add some lines to `maven/src/changes/changes.xml` to describe the changes you've made?

brandonherzog commented:
Ok, I rebased onto your latest to grab your XER fix and updated the changelog from the commits. 

Also made one fix because apparently XER files and the P6 database seem to store the scheduling options differently and in different tables.

Passes all the tests on my project, so I'm happy with how the code stands.
